### DISCUSSION TITLE
why useMutation onError never gets fired even when the server is down or API responds with 500?

### DISCUSSION THREAD

**Discussion by ritikbanger** [TanStack/query#3931] [Q&A]
Version:   "react-query": "^3.39.1",

I am using useMutation hook and calling the mutateAsync function. The onSuccess of mutationAync fires always even when the API does not respond or responds with error code 500 or it fails.

Why onError is not triggered in such cases, is there a need to pass something explicitly somewhere or am I missing something here.

By defination of onError: This function will fire if the mutation encounters an error and will be passed the error.
If a promise is returned, it will be awaited and resolved before proceeding.

Since, if the api fails, the mutation will not receive any results or there will be no response but still onError is not trigerred.
Code:

```
  const { mutateAsync } = useMutation(
    (params: { some Param goes here}) => {
      return backendServiceCall(params)
    }
  )
<!-- some code goes here -->
 const functionThatCallsTheMutateAsync = async () => {
    await mutateAsync(
      { params },
      {
        onSuccess: () => {
          sendMessage('success')
        },
        onError: () => sendMessage('error' ),
      }
    )
  }

functionThatCallsTheMutateAsync()

```
```
Case: API Failure
Output: success
```

**Comment by TkDodo** [comment]
> Since, if the api fails, the mutation will not receive any results or there will be no response but still onError is not trigerred.

you are not showing your api function :/ very likely, it doesn't return a rejected promise in case of an "error code 500". For example, the browser native `fetch` API will behave this way...

**Reply by ritikbanger** [reply]
![Screenshot from 2022-07-28 17-39-02](https://user-images.githubusercontent.com/47841501/181501458-d5db3d80-7e35-44c2-a152-853f854038cd.png)

I am looking for onError trigger on such cases when the backend server is not running or is down.

**Reply by TkDodo** [reply]
you're still not showing _how_ you make the request!

**Reply by ritikbanger** [reply]
I am using axios.

```
export const someApiService = (params): Promise<ApiResponse> => {
  return http.post(endpoint.someEndpoint, data)
}
```

The http method:

```
export const http = axios.create({
  baseURL: config.baseUrl,
  headers: { 'Content-Type': 'application/json' },
})

http.interceptors.request.use((req) => {
  req.headers.authorization = `Bearer Token`

  return req
})

http.interceptors.response.use(
  (res) => {
    return withData(res.data) as AxiosResponse<any>
  },
  (err) => {
    return withError(err?.response?.data?.error)
  }
)

export const httpAuth = axios.create({
  baseURL: config.loginUrl,
  headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
})

httpAuth.interceptors.request.use((req) => {
  return req
})

export function post<D, P>(url: string, data: D, params?: P): any {
  return http({
    method: 'post',
    url,
    data,
    params,
  })
}
```

**Reply by TkDodo** [reply]
i guess the interceptor does it? what is `withError`? (you somehow manage to always leave out the important parts, which is why I usually request a running reproduction that shows the issue!)

as you can see in the [axios interceptor docs](https://axios-http.com/docs/interceptors), you need to return a rejected promise from the error interceptor. Otherwise, it likely transforms it to a resolved promise.

anyways, as long as react-query gets a rejected promise, it will treat those as errors. somewhere in your chain, you get rid of the rejected promise and transform it to a resolved promise. that's what you need to find.

**Reply by ritikbanger** [reply]
Sorry for the missing withError.

Here is some more information around withError:

```
const toString = Object.prototype.toString

export const isObject = <T>(arg: T): boolean => {
  return toString.call(arg) === '[object Object]'
}

export const withError = <T>(arg: T): ApiResponse => {
  if (isObject(arg)) {
    return {
      data: null,
      error: {
        ...arg,
      },
    }
  }

  return {
    data: null,
    error: {
      message: arg,
    },
  }
}

export const withData = <T>(data: T): ApiResponse => ({
  error: null,
  data,
})

```

Here is the ApiResponse interface.


```
export type TApiState = Record<string, any> | null

export default interface ApiResponse<T = any> extends Partial<AxiosResponse<T | TApiState>> {
  data: TApiState
  error: TApiState
}

```

Let me know if you need anything else for the context.

**Reply by ritikbanger** [reply]
Any update on this @TkDodo

**Reply by TkDodo** [reply]
if you want me to look further into this, at least provide a codesandbox reproduction that shows your issue please

**Reply by ritikbanger** [reply]
Without backend, how can I create a codsandbox here.

**Reply by TkDodo** [reply]
You have the query function return Promise.resolve directly with some json data?

**Reply by ritikbanger** [reply]
Yes. I respond with the JSON data and the status.

**Comment by acharyayush** [comment]
yeah same problem.

```
//mutation hook
const { data, mutate: mutateCustomer } = useMutation({
    mutationFn: createCustomer,
    onSuccess: () => {
      console.log("navigated")
      navigate("/")
    },
    onError: () => {
      console.log("error occured")
    },
  })
//api.js
//i have BASEURL
export const createCustomer = async (customerDetail) => {
  try {
    const { data } = await axios.post(
      `${BASEURL}/api/65865d8b75a727705c869123/customers/add`,
      customerDetail
    )
    toast("success", "New customer is created 😁")
    return data
  } catch (err) {
    toast("error", err.message)
    return err
  }
}
```
if i submit with empty customer details, i get internal server error but onSuccess gets triggered and page is navigated to homepage, onError never gets triggerred...

**Reply by TkDodo** [reply]
```diff
} catch (err) {
toast("error", err.message)
- return err
+ throw err
}
```

sorry but y'all need to learn how promises work, or at least read my blog because I've answered this exact question so many times that I've put it in my FAQ a long time ago:

- https://tkdodo.eu/blog/react-query-fa-qs#logging

**Reply by acharyayush** [reply]
Thank you for answering and for the blog.

**Reply by deeseeker** [reply]
This came in very handy for me in my project. It took me a while before I found this but very helpful. Long live TkDodo

**Reply by hokwanhung** [reply]
This solved one of my biggest questions about `useMutation`. Thank you @TkDodo!
